src/DotNetty.Codecs/Base64/Base64.cs

```diff
@@ -0,0 +1,374 @@
+            {
+                if (breakLines)
+                {
+                    if (charCount == MAX_LINE_LENGTH)
+                    {
+                        dest.SetByte(j++, NEW_LINE);
+                        charCount = 0;
+                    }
+                    charCount += 4;
+                }
+
+                dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                dest.SetByte(j + 1, alphabet[((src.GetByte(i) & 0x03) << 4) | ((src.GetByte(i + 1) & 0xf0) >> 4)]);
+                dest.SetByte(j + 2, alphabet[((src.GetByte(i + 1) & 0x0f) << 2) | ((src.GetByte(i + 2) & 0xc0) >> 6)]);
+                dest.SetByte(j + 3, alphabet[(src.GetByte(i + 2) & 0x3f)]);
+                j += 4;
+            }
+
+            i = calcLength;
+
+            if (breakLines && (length % 3 != 0) && (charCount == MAX_LINE_LENGTH))
+            {
+                dest.SetByte(j++, NEW_LINE);
+            }
+            switch (length % 3)
+            {
+                case 2:
+                    dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                    dest.SetByte(j + 1, alphabet[((src.GetByte(i) & 0x03) << 4) | ((src.GetByte(i + 1) & 0xf0) >> 4)]);
+                    dest.SetByte(j + 2, alphabet[(src.GetByte(i + 1) & 0x0f) << 2]);
+                    dest.SetByte(j + 3, EQUALS_SIGN);
+                    j += 4;
+                    break;
+                case 1:
+                    dest.SetByte(j + 0, alphabet[(src.GetByte(i) & 0xfc) >> 2]);
+                    dest.SetByte(j + 1, alphabet[(src.GetByte(i) & 0x03) << 4]);
+                    dest.SetByte(j + 2, EQUALS_SIGN);
+                    dest.SetByte(j + 3, EQUALS_SIGN);
+                    j += 4;
+                    break;
+            }
+            //restore src's ReaderIndex and WriterIndex
+            src.SetIndex(srcReaderIndex, srcWriterIndex);
+            //remove last byte if it's NewLine
+            int destLength = dest.GetByte(j - 1) == NEW_LINE ? j - 1 : j;
+            return destLength;
+        }
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator allocator)
+        {
+            if (src == null)
+            {
+                throw new ArgumentNullException(nameof(src));
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException(nameof(dialect.alphabet));
+            }
+            Contract.Assert(dialect.alphabet.Length == 64, "alphabet.Length must be 64!");
+            if (offset < src.ReaderIndex || offset + length > src.ReaderIndex + src.ReadableBytes)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            if (length <= 0)
+            {
+                return Unpooled.Empty;
+            }
+
+            int outLength = length / 3 * 4;
```

Comment:
i realize what you r talking about.  netty use a temp variable named `e` , which is in Base64.java and line 128(netty 4.1 release).
netty's impl means it allocate a little bigger more space, more than what it really need.
in fact, i read both .net framework and netty's base64 encoding/decoding. i think .net's encoding is  easy to understand, and its decoding is more complex and difficult to understand.

https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/src/mscorlib/src/System/Convert.cs#L2341